kv/dist_sender.go

```diff
@@ -527,12 +577,25 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
 // to be handled separately by the caller.
 func (ds *DistSender) sendAttempt(trace *tracer.Trace, args proto.Request, reply proto.Response, desc *proto.RangeDescriptor) error {
 	defer trace.Epoch("sending RPC")()
-	// Truncate the request to our current range, making sure not to
-	// touch it unless we have to (it is illegal to send EndKey on
-	// commands which do not operate on ranges).
-	if endKey := args.Header().EndKey; endKey != nil && !endKey.Less(desc.EndKey) {
-		defer func(k proto.Key) { args.Header().EndKey = k }(endKey)
-		args.Header().EndKey = desc.EndKey
+	if revScanArgs, ok := args.(*proto.ReverseScanRequest); ok {
+		// Truncate the request to our current range, making sure not to
+		// touch it unless we have to (it is illegal to send key on
+		// commands which do not operate on ranges).
+		if key := args.Header().Key; key != nil && key.Less(desc.StartKey) {
+			defer func(k proto.Key) { args.Header().Key = k }(key)
+			args.Header().Key = desc.StartKey
+			revScanArgs.InclusiveKey = true
```

Comment:
seems like `InclusiveKey` is tuned towards this multi-range case, but if it's there it should probably also work for scans on a single range. Currently it doesn't because you overwrite it below. Maybe it would be better if the semantics for `ReverseScan` would be that it excludes the `EndKey`, but always includes `StartKey`? Then you have less trouble, and clients that want the opposite behaviour can simply scan from `EndKey.Next()` to `StartKey.Next()`.